[wip] towards compiling with assumptions

o- opened the pull request:
based on #272 (ignore first two commits)

I open this to start a discussion how the api of assumptions could look like.

The idea is that there is an *OptimizationContext* under which code is compiled. The OptimizationContext captures the things the compiler is allowed to take for granted.

Then the next question will be, how does the runtime ensure the context is always matching. But in this PR I only use the context for things we can deduce from contextual information.

## WIP

There is a problem here, namely we store all versions into slot 1 and we always call slot 1, even functions compiled under assumptions.

Here:
https://github.com/reactorlabs/rir/blob/master/rir/src/compiler/translations/pir_2_rir.cpp#L955 
we would potentially end up compiling a function compiled under assumptions into the generic slot1.

To fix this my first impulse woud be to just extend `static_call_` instruction with a fixed slot index.
That should be enough for the case here. Otoh, should we already start to tie it in with the signature mechanism? @mhyee any opinions?

o- committed:
massive cleanup of scope analysis

o- committed:
fix merging of two pir abstract values

when merging the type (which is supposed to be the union type
of all possible values) was not updated.

o- committed:
versioned closures are not used anywhere

o- committed:
fix a very old bug, calling call_static_ with wrong num of args

call_static_ is only valid if the number of arguments exactly matches!
in case of mismatching arguments we need to supply R_MissingValue,
but that is not yet supported in pir.

o- committed:
towards compiling with assumptions

We introduce a compilation context that contains the assumptions
we compile under and the environment the function is closed over.

Currently this is just used to encode contextual "static" knowledge,
but it introduces the infrastructure, that we could start using
for dynamic assumptions.

o- committed:
make this sound, but stupid

o- commented:
To understand the proposal just look at  ....

Update: my `git add -p` skills are terrible. Had to fix it. The main commit is now 1bf12697f0455472f1e1df66c9da3fe8a2855b2d

mhyee commented on rir/src/compiler/pir/optimization_context.h:
```diff
@@ -0,0 +1,43 @@
+    OptimizationContext(Env* environment, const AssumptionsSet& assumptions)
+        : environemnt(environment), assumptions(assumptions) {}
+
+    Env* environemnt;
```
environemnt -> environment

mhyee commented:
> Then the next question will be, how does the runtime ensure the context is always matching. But in this PR I only use the context for things we can deduce from contextual information.

What's the joke about how "context" is used in computer science for whatever stuff we want to keep around but don't know what to name it? More seriously, "use the context for things we can deduce from contextual information" seems like a recursive definition.

> [...]
> To fix this my first impulse woud be to just extend `static_call_` instruction with a fixed slot index.
> That should be enough for the case here. Otoh, should we already start to tie it in with the signature mechanism? @mhyee any opinions?

I don't think we have a good sense of how to tie the assumptions with the signatures? It's all very uncertain right now, so I would go with the simplest solution for now, and expect to change it again (and again) later.

o- committed:
dynamic dispatch based on signature

in this commit we order versions in the dispatch table by increasing
number of assumptions. To dispatch we search backwards through the
list. Entry 0, the baseline, always matches.

The assumptions are just a static bitset. Probably we want something
more elaborate, but since we are not exactly sure what we want, it's
probably a good first iteration.

o- committed:
implement static arguments matching

support static calls with named arguments in pir, by statically matching
named arguments to their position and reordering at compile time.

charig commented on rir/src/runtime/DispatchTable.h:
```diff
@@ -18,30 +18,56 @@ typedef SEXP DispatchTableEntry;
+               FunctionSignature::BaselineVersion);
+        setEntry(0, f->container());
+        if (size() == 0)
+            size_++;
```
We are assuming there will always be a baseline compiled. Maybe that is not necessary...

charig commented on rir/src/runtime/FunctionSignature.h:
```diff
@@ -2,13 +2,25 @@
 namespace rir {
 
 struct FunctionSignature {
+    enum EnvironmentCreation {
```
Environment
 Provided
 NotProvided

charig commented on rir/src/runtime/FunctionSignature.h:
```diff
@@ -35,50 +47,60 @@ struct FunctionSignature {
+        }
     }
 
     FunctionSignature() = default;
```
Shouldn't we remove the default constructor now?

charig commented on rir/src/runtime/DispatchTable.cpp:
```diff
@@ -0,0 +1,3 @@
+#include "DispatchTable.h"
```
?

charig commented on rir/src/compiler/pir/assumptions.h:
```diff
@@ -0,0 +1,32 @@
+    LAST = CorrectOrderOfArguments
+};
+
+struct AssumptionsSet : public EnumSet<Assumptions> {
```
Maybe enum is an Assumption  and here we can call it Assumptions
